x-pack/solutions/observability/plugins/apm/server/routes/diagnostics/service_map/get_exit_spans_from_samples.ts

```diff
@@ -0,0 +1,236 @@
+  start,
+  end,
+  destinationNode,
+  ids,
+}: {
+  apmEventClient: APMEventClient;
+  start: number;
+  end: number;
+  sourceNode: string;
+  destinationNode: string;
+  ids: string[];
+}) {
+  const response = await apmEventClient.search('diagnostics_get_exit_spans_from_source_node', {
+    apm: {
+      events: [ProcessorEvent.transaction],
+    },
+    track_total_hits: false,
+    size: 0,
+    query: {
+      bool: {
+        filter: [...rangeQuery(start, end), ...termsQuery(PARENT_ID, ...ids)],
+      },
+    },
+    aggs: {
+      matching_destination_resources: {
+        filter: {
+          term: {
+            [SERVICE_NAME]: destinationNode,
+          },
+        },
+        aggs: {
+          sample_docs: {
+            top_hits: {
+              size: 5,
+            },
+          },
+        },
+      },
+      destination_services: {
+        terms: {
+          field: SERVICE_NAME,
+          size: 50,
+        },
+        aggs: {
+          sample_docs: {
+            top_hits: {
+              size: 5,
+            },
+          },
+        },
+      },
+    },
+  });
+
+  const apmExitSpans =
+    response?.aggregations?.destination_services?.buckets?.map((item: any) => {
+      const doc = item?.sample_docs?.hits?.hits?.[0]?._source;
+      return {
+        destinationService: doc?.service?.name ?? '',
+        spanSubType: doc?.span?.subtype ?? '',
+        spanId: doc?.span?.id ?? '',
+        spanType: doc?.span?.type ?? '',
+        transactionId: doc?.transaction?.id ?? '',
+        serviceNodeName: doc?.service?.node?.name ?? '',
+        traceId: doc?.trace?.id ?? '',
+        agentName: doc?.agent?.name ?? '',
+        docCount: item?.doc_count ?? 0,
+        isOtel: false,
+      };
+    }) || [];
+
+  const matchingCount =
+    response?.aggregations?.matching_destination_resources?.sample_docs?.hits?.total?.value || 0;
+  return {
+    apmExitSpans,
+    totalConnections: apmExitSpans.length,
+    rawResponse: response,
+    hasMatchingDestinationResources: matchingCount > 0,
+  };
+}
+
+export async function getSourceSpanIds({
+  apmEventClient,
+  start,
+  end,
+  sourceNode,
+  traceIds,
+}: {
+  apmEventClient: APMEventClient;
+  start: number;
+  end: number;
+  sourceNode: string;
+  traceIds: string[];
+}) {
+  const response = await apmEventClient.search('diagnostics_get_source_node_span_samples', {
+    apm: {
+      events: [ProcessorEvent.span],
+    },
+    track_total_hits: false,
+    size: 0,
+    query: {
+      bool: {
+        filter: [
+          ...rangeQuery(start, end),
+          ...termsQuery(SERVICE_NAME, sourceNode),
+          ...termsQuery(TRACE_ID, ...traceIds),
+        ],
+      },
+    },
+    aggs: {
+      sample_docs: {
+        terms: {
+          field: SPAN_NAME,
+          size: 500,
+        },
+        aggs: {
+          top_span_ids: {
+            top_hits: {
+              size: 10,
+              _source: {
+                include: [SPAN_ID],
+              },
```

Comment:
I had a problem with `_source` and otel: https://github.com/elastic/kibana/pull/229487/files#diff-c8025d14a10402a46052d3c6dfde073aa4be21d8578ec763098c4f1de3beb980.

Reply:
good catch. Updated it here https://github.com/elastic/kibana/pull/228795/commits/6cc097a4629d070a6f2274fcc93515b4c37674e5